### DISCUSSION TITLE
Best practice to have persistence in OPC UA Server

### DISCUSSION THREAD

**Discussion by francescobianca** [FreeOpcUa/opcua-asyncio#788] [Q&A]
Hi everyone,

I have my OPCUA server where at startup I do the xml import with the semantic model and then load the various nodes via a json file. I would like to know what are the best practices to keep the semantic model updated (for example if via api I add a node I want to know how to make a persistent backup). Maybe through this library there is already something supported (I don't know a mongodb, a sql etc.)

**Comment by AndreasHeine** [comment]
if you want to persist nodes which have been created while runtime, you need to export them to XML!? or as shelffile!
but after re-import you need to link them logically if they have some behavior attached.

https://github.com/FreeOpcUa/opcua-asyncio/blob/c75f45d518e48daacd638755e4085f311ad673db/asyncua/server/server.py#L587

https://github.com/FreeOpcUa/opcua-asyncio/blob/c75f45d518e48daacd638755e4085f311ad673db/asyncua/server/server.py#L595

https://github.com/FreeOpcUa/opcua-asyncio/blob/c75f45d518e48daacd638755e4085f311ad673db/asyncua/server/server.py#L107
